src/lattice.cpp

```diff
@@ -710,69 +710,87 @@ const
 std::array<int, 3>
 HexLattice::get_indices(Position r, Direction u) const
 {
-  // The implementation for HexLattice currently doesn't use direction
-  // information. As a result, we move the position slightly forward to
-  // determine what lattice index the particle is most likely to be in.
-  r += TINY_BIT * u;
+  // result variables
+  int ix{}, ia{}, iz{};
 
   // Offset the xyz by the lattice center.
   Position r_o {r.x - center_.x, r.y - center_.y, r.z};
   if (is_3d_) {r_o.z -= center_.z;}
 
-  // Index the z direction.
-  std::array<int, 3> out;
+  // Index the z direction. Same as the technique used in
+  // RectLattice::get_indices.
   if (is_3d_) {
-    out[2] = std::floor(r_o.z / pitch_[1] + 0.5 * n_axial_);
-  } else {
-    out[2] = 0;
+    double iz_ {r_o.z / pitch_[1] + 0.5 * n_axial_};
+    long iz_close {std::lround(iz_)};
+    if (std::abs(iz_ - iz_close) < FP_COINCIDENT) {
```

Comment:
Might as well use your nifty `coincident` function:
```C++
if (coincident(iz_, iz_close)) {
```
`iz_close` will get converted to a double either way